zcash_primitives/src/transaction/components/sapling.rs

```diff
@@ -40,49 +111,78 @@ impl std::fmt::Debug for SpendDescription {
-            bls12_381::Scalar::from_repr(f)
-                .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "anchor not in field"))?
-        };
+/// Consensus rules (§7.3) & (§7.4):
+/// - Canonical encoding is enforced here
+pub fn read_scalar<R: Read>(mut reader: R, field: &str) -> io::Result<bls12_381::Scalar> {
+    let mut f = [0u8; 32];
+    reader.read_exact(&mut f)?;
+    bls12_381::Scalar::from_repr(f).ok_or_else(|| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            field.to_owned() + " not in field",
+        )
+    })
+}
+
+/// Consensus rules (§4.4) & (§4.5):
+/// - Canonical encoding is enforced by the API of SaplingVerificationContext::check_spend()
+///   and SaplingVerificationContext::check_output() due to the need to parse this into a
+///   bellman::groth16::Proof.
+/// - Proof validity is enforced in SaplingVerificationContext::check_spend()
+///   and SaplingVerificationContext::check_output()
+pub fn read_zkproof<R: Read>(mut reader: R) -> io::Result<GrothProofBytes> {
+    let mut zkproof = [0u8; GROTH_PROOF_SIZE];
+    reader.read_exact(&mut zkproof)?;
+    Ok(zkproof)
+}
 
+impl SpendDescription<Authorized> {
+    pub fn read_nullifier<R: Read>(mut reader: R) -> io::Result<Nullifier> {
         let mut nullifier = Nullifier([0u8; 32]);
         reader.read_exact(&mut nullifier.0)?;
+        Ok(nullifier)
+    }
 
-        // Consensus rules (§4.4):
-        // - Canonical encoding is enforced here.
-        // - "Not small order" is enforced in SaplingVerificationContext::check_spend()
-        let rk = PublicKey::read(&mut reader)?;
+    /// Consensus rules (§4.4):
+    /// - Canonical encoding is enforced here.
+    /// - "Not small order" is enforced in SaplingVerificationContext::check_spend()
+    pub fn read_rk<R: Read>(mut reader: R) -> io::Result<PublicKey> {
+        PublicKey::read(&mut reader)
+    }
 
-        // Consensus rules (§4.4):
-        // - Canonical encoding is enforced by the API of SaplingVerificationContext::check_spend()
-        //   due to the need to parse this into a bellman::groth16::Proof.
-        // - Proof validity is enforced in SaplingVerificationContext::check_spend()
-        let mut zkproof = [0u8; GROTH_PROOF_SIZE];
-        reader.read_exact(&mut zkproof)?;
+    /// Consensus rules (§4.4):
+    /// - Canonical encoding is enforced here.
+    /// - Signature validity is enforced in SaplingVerificationContext::check_spend()
+    pub fn read_spend_auth_sig<R: Read>(mut reader: R) -> io::Result<Signature> {
+        Signature::read(&mut reader)
+    }
 
-        // Consensus rules (§4.4):
-        // - Canonical encoding is enforced here.
-        // - Signature validity is enforced in SaplingVerificationContext::check_spend()
-        let spend_auth_sig = Some(Signature::read(&mut reader)?);
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let cv = read_point(&mut reader, "cv")?;
```

Comment:
I think I'd prefer to keep the consensus rule comments here, since they apply to the specific variable types, not to the general `read_point` etc. functions (even though it is those functions that enforce the rules enforced "here").